src/Git.hs

```diff
@@ -201,100 +199,99 @@ instance FromJSON GitIntegrationFailure
 
 instance ToJSON GitIntegrationFailure where toEncoding = Aeson.genericToEncoding Aeson.defaultOptions
 
-data GitOperationFree a
-  = FetchBranch Branch FetchWithTags a
-  | ForcePush Sha Branch (PushResult -> a)
-  | Push Sha Branch (PushResult -> a)
-  | PushAtomic [SomeRefSpec] (PushResult -> a)
-  | Rebase Sha RemoteBranch (Maybe Sha -> a)
-  | Merge Sha Text (Maybe Sha -> a)
-  | Checkout RemoteBranch (Maybe Sha -> a)
-  | Clone RemoteUrl (CloneResult -> a)
-  | GetParent Sha (Maybe Sha -> a)
-  | RevParse SomeRefSpec (Maybe Sha -> a)
-  | DoesGitDirectoryExist (Bool -> a)
-  | LastTag Sha (Maybe Text -> a)
-  | ShortLog SomeRefSpec SomeRefSpec (Maybe Text -> a)
-  | Tag Sha TagName TagMessage (TagResult -> a)
-  | DeleteTag TagName a
-  | DeleteBranch Branch a
-  | DeleteRemoteBranch Branch (PushResult -> a)
-  | CheckOrphanFixups Sha RemoteBranch (Bool -> a)
-  deriving (Functor)
-
-type GitOperation = Free GitOperationFree
-
-fetchBranch :: Branch -> GitOperation ()
-fetchBranch remoteBranch = liftF $ FetchBranch remoteBranch NoTags ()
-
-fetchBranchWithTags :: Branch -> GitOperation ()
-fetchBranchWithTags remoteBranch = liftF $ FetchBranch remoteBranch WithTags ()
-
-forcePush :: Sha -> Branch -> GitOperation PushResult
-forcePush sha remoteBranch = liftF $ ForcePush sha remoteBranch id
-
-push :: Sha -> Branch -> GitOperation PushResult
-push sha remoteBranch = liftF $ Push sha remoteBranch id
-
-pushAtomic :: [SomeRefSpec] -> GitOperation PushResult
-pushAtomic refs = liftF $ PushAtomic refs id
-
-rebase :: Sha -> RemoteBranch -> GitOperation (Maybe Sha)
-rebase sha ontoBranch = liftF $ Rebase sha ontoBranch id
-
-merge :: Sha -> Text -> GitOperation (Maybe Sha)
-merge sha message = liftF $ Merge sha message id
+data GitOperation :: Effect where
+  FetchBranch :: Branch -> FetchWithTags -> GitOperation m ()
+  ForcePush :: Sha -> Branch -> GitOperation m PushResult
+  Push :: Sha -> Branch -> GitOperation m PushResult
+  PushAtomic :: [SomeRefSpec] -> GitOperation m PushResult
+  Rebase :: Sha -> RemoteBranch -> GitOperation m (Maybe Sha)
+  Merge :: Sha -> Text -> GitOperation m (Maybe Sha)
+  Checkout :: RemoteBranch -> GitOperation m (Maybe Sha)
+  Clone :: RemoteUrl -> GitOperation m CloneResult
+  GetParent :: Sha -> GitOperation m (Maybe Sha)
+  RevParse :: SomeRefSpec -> GitOperation m (Maybe Sha)
+  DoesGitDirectoryExist :: GitOperation m Bool
+  LastTag :: Sha -> GitOperation m (Maybe Text)
+  ShortLog :: SomeRefSpec -> SomeRefSpec -> GitOperation m (Maybe Text)
+  Tag :: Sha -> TagName -> TagMessage -> GitOperation m TagResult
+  DeleteTag :: TagName -> GitOperation m ()
+  DeleteBranch :: Branch -> GitOperation m ()
+  DeleteRemoteBranch :: Branch -> GitOperation m PushResult
+  CheckOrphanFixups :: Sha -> RemoteBranch -> GitOperation m Bool
+
+type instance DispatchOf GitOperation = 'Dynamic
```

Comment:
"Dynamic", as opposed to Static dispatch is an `effectful` concept. It means that there can be multiple implementations (called `interpretations`) for the above action. With Static dispatch, there is always precisely one interpretation.

The reason the difference exists, I believe, is mostly performance. Static is a little faster, though Dynamic is still very fast.